File: Low Level Design (LLD)/DesignPatterns/03. Singleton/CPP/ByPointer/Logger.cpp
#include "./Logger.h"

int Logger::counter = 0;
Logger* Logger::loggerInstance = nullptr;
std::mutex Logger::mtx;

Logger::Logger() {
    counter++;
    std::cout << "Instance created, no of instance: " << counter << std::endl;
}

Logger* Logger::getLogger() {
    if (loggerInstance == nullptr) {
        mtx.lock();
        if (loggerInstance == nullptr) {
            loggerInstance = new Logger();
        }
        mtx.unlock();
    }
    return loggerInstance;
}

void Logger::Log(std::string message) {
    std::cout << message << std::endl;
}

File: Low Level Design (LLD)/DesignPatterns/01. Creational/Singleton/Java/Bill Pugh Method/Logger.java
public class Logger {

    private static int counter = 0;

    private Logger() {
        counter++;
        System.out.println("Instance created, number of instances: " + counter);
    }

    private static class LoggerHelper {
        private static final Logger INSTANCE = new Logger();
    }

    public static Logger getLogger() {
        return LoggerHelper.INSTANCE;
    }

    public void log(String message) {
        System.out.println(message);
    }
}
